frame/benchmarking/README.md

```diff
@@ -1,3 +1,182 @@
+
+In general, benchmarks returning `Ok(())` is all you need to check for since it signals the executed
+extrinsic has completed successfully. However, you can optionally include a `verify` block with your
+benchmark which can additionally verify any final conditions, such as the final state of your
+runtime.
+
+These additional `verify` blocks will not effect the results of your final benchmarking process.
+
+To run the tests, you need to be sure to enable the `runtime-benchmarks` feature flag. This may also
+mean you need to move into your node's binary folder. For example, with the Substrate repository,
+this is how you would test the Balances Pallet's benchmarks:
+
+```bash
+cd bin/node/cli
+cargo test -p pallet-balances --features runtime-benchmarks
+```
+
+## Adding Benchmarks
+
+The benchmarks included with each pallet are not automatically added to your node. To actually
+execute these benchmarks, you need to implement the `frame_benchmarking::Benchmark` trait. You can
+see an example of how to do this in the [included Substrate
+node](../../bin/node/runtime/src/lib.rs).
+
+Assuming there are already some benchmarks set up on your node, you just need to add another
+instance of the `add_benchmarks!` macro:
+
+```rust
+///  configuration for running benchmarks
+///               |    name of your pallet's crate (as imported)
+///               v                   v
+add_benchmark!(params, batches, pallet_balances, Balances);
+///                       ^                          ^
+///    where all benchmark results are saved         |
+///            the `struct` created for your pallet by `construct_runtime!`
+```
+
+Once you have done this, you will need to compile your node binary with the `runtime-benchmarks`
+feature flag:
+
+```bash
+cd bin/node/cli
+cargo build --release --features runtime-benchmarks
+```
+
+## Running Benchmarks
+
+Finally, once you have a node binary with benchmarks enabled, you need to execute your various
+benchmarks.
+
+You can get a list of the available benchmarks by running:
+
+```bash
+./target/release/substrate benchmark --chain dev --pallet "*" --extrinsic "*" --repeat 0
+```
+
+Then you can run a benchmark like so:
+
+```bash
+./target/release/substrate benchmark \
+    --chain dev \               # Configurable Chain Spec
+    --execution=wasm \          # Always test with Wasm
+    --wasm-execution=compiled \ # Always used `wasm-time`
+    --pallet pallet_balances \  # Select the pallet
+    --extrinsic transfer \      # Select the extrinsic
+    --steps 50 \                # Number of samples across component ranges
+    --repeat 20 \               # Number of times we repeat a benchmark
+    --output \                  # Output benchmark results into a Rust file
+```
+
+This will output a file `pallet_name.rs` which implements the `WeightInfo` trait you should include
+in your pallet. Each blockchain should generate their own benchmark file with their custom
```

Comment:
This is a bit unclear to me: the file is created in the crate root? Need to be copied into the runtime source file? Or…?

Do we have any tooling available to list un-weighted pallets or dispatchables? Like "test coverage" but for weights.

Reply:
file will be created in your working directory where you executed the command. It will be a todo feature to support a custom path for the file.

No tools available atm for test coverage of weights. also a good idea for CI